distribution/lib/Standard/Database/0.0.0-dev/src/Data/Column.enso

```diff
@@ -550,26 +559,136 @@ type Column
+         multiple of 10^(-n)". For negative decimal counts, this results in
+         rounding to the nearest positive integer power of 10.
+
+       > Example
+         Round a column to two decimal places.
+
+             import Standard.Examples
+
+             example_round = Examples.decimal_column.round 2
+    round : Integer -> Boolean -> Column | Illegal_Argument | Invalid_Value_Type
+    round self decimal_places=0 use_bankers=False =
+        Decimal_Places_Helpers.check_decimal_places decimal_places <| Value_Type.expect_numeric self <|
+            new_name = self.naming_helpers.function_name "round" [self]
+            scale = 10 ^ decimal_places
+            scaled = self * scale
+            round_base = scaled.floor
+            round_midpoint = (round_base + 0.5) / scale
+            even_is_up = (self >= 0).iif ((scaled.truncate % 2) != 0) ((scaled.truncate % 2) == 0)
+            half_goes_up = if use_bankers then even_is_up else (self.const True)
+            do_round_up = half_goes_up.iif (self >= round_midpoint) (self > round_midpoint)
+            result_float = do_round_up.iif ((round_base + 1.0) / scale) (round_base / scale)
 
-    ## Truncating values is not supported in database columns.
+            should_cast_to_integer = decimal_places <= 0 || self.value_type.is_integer
+            result = if should_cast_to_integer.not then result_float else
+                result_float.cast Value_Type.Integer
+
+            result.rename new_name
+
+    ## ALIAS int
+       UNSTABLE
+
+       Truncate the floating-point values to an integer by dropping the
+       fractional part. This is equivalent to "round-toward-zero".
+
+       > Example
+         Truncate a column of floating-point values.
+
+             import Standard.Examples
+
+             example_truncate = Examples.decimal_column.truncate
     truncate : Column ! Invalid_Value_Type
     truncate self =
-        Error.throw <| Unsupported_Database_Operation.Error "`Column.truncate` is not implemented yet for the Database backends."
+        Value_Type.expect_numeric self <|
+            new_name = self.naming_helpers.function_name "truncate" [self]
+            self.make_unary_op "TRUNCATE" new_name
+
+    ## UNSTABLE
+
+       Takes the ceiling of floating-point values, returning integer values.
 
-    ## Taking the ceiling of values is not supported in database columns.
+       > Example
+         Take the ceiling of a column of floating-point values.
+
+             import Standard.Examples
+
+             example_truncate = Examples.decimal_column.ceil
     ceil : Column ! Invalid_Value_Type
     ceil self =
-        Error.throw <| Unsupported_Database_Operation.Error "`Column.ceil` is not implemented yet for the Database backends."
+        Value_Type.expect_numeric self <|
+            new_name = self.naming_helpers.function_name "ceil" [self]
+            self.make_unary_op "CEIL" new_name
```

Comment:
Looking at this + Base_Generator impls of ops, I realised that if I have a column of `Integer 32-bit` and I `floor` it, it will become a column of `Integer 64-bit` (because you cast to `BIGINT` in the op implementation). That widening seems unnecessary, right?

The `BIGINT` cast itself seems ok after a thought - when casting from floating-point we want to ensure the result is an integer value and defaulting to 64-bit ints seems sensible (1).

IMO we should check _here_ if the column is already and integer - and for floor/ceil we can then just return the current expression unmodified (if its already an int, no need to call the operation - which may actually cause a conversion to float and loss of precision on big integers!).

> 1. My only worry here is that we may overflow when flooring a decimal. Actually, should we not ensure that `Decimal` after floor/ceil/round remains a `Decimal`? After all if these are arbitrary precision number, it could have been a number exceeding the range of 64-bit ints, so we may get wrong results if we always cast to ints. IMO flooring/rounding a Decimal should still be yielding a Decimal. What do you think?

Reply:
To sum up, to ensure correctness:
1. I'd not perform ceil/floor/trunc if the input column is `self.value_type.is_integer` - there is no work to be done. We can rename the column to have a consistent name, but other than that it should act as identity.
	- We don't want to risk that if we start with `MAX_LONG`, possibly converting it into a float for the `floor` operation may lose precision and coerce it to some closest representable floating point number which after casting back to `int64` may no longer equal `MAX_LONG` (due to floats not being able to represent exactly all integers > 2**52.
	- Ideally if we do `round decimal_places>=0` it should also do the same.
2. I think all operations ceil/floor/trunc/round - should return back a Decimal when given a Decimal at input.

And ideally it could be good if we can add some tests for these scenarios. Sth like these:

```
Test.specify "should not lose precision rounding large integers" <|
    i1 = Long.MAX_VALUE - 1
    i2 = i1 - 1
    t = table_builder [["X", [i1, i2]]]
    [(.floor), (.ceil), (.trunc), (x-> x.round 0), (x-> x.round 2)].each op->
        op (t.at "X") . to_vector . should_equal [i1, i2]
if setup.test_selection.supports_decimal_type then
	Test.specify "should return decimals when rounding decimals" <|
    i1 = Long.MAX_VALUE - 1
    c = table_builder [["X", [i1]]] . at "X"
    decimal_col = c.cast Value_Type.Decimal
    decimal_col.value_type.is_decimal . should_be_true
    decimal_col2 = decimal_col + decimal_col*decimal_col
    [(.floor), (.ceil), (.trunc), (x-> x.round 0), (x-> x.round 2)].each op->
        op decimal_col2 . to_vector . should_equal [i1 + i1*i1]
```

Reply:
I agree an integer type should return the same type.
Likewise Decimal should not change its type.

Reply:
As discussed, we're conforming to the backend's return types.